dm/syncer/binlogstream/stream_modifier.go

```diff
@@ -0,0 +1,252 @@
+func (m *streamModifier) Set(req *pb.HandleWorkerErrorRequest, events []*replication.BinlogEvent) error {
+	// precheck
+	switch req.Op {
+	case pb.ErrorOp_Skip:
+		if len(events) != 0 {
+			m.logger.Warn("skip op should not have events", zap.Int("eventLen", len(events)))
+		}
+	case pb.ErrorOp_Replace, pb.ErrorOp_Inject:
+		if len(events) == 0 {
+			return terror.ErrSyncerEvent.Generatef("%s op should have non-empty events", req.Op.String())
+		}
+	default:
+		m.logger.DPanic("invalid error op", zap.String("op", req.Op.String()))
+		return terror.ErrSyncerEvent.Generatef("invalid error op: %s", req.Op.String())
+	}
+
+	pos, err := binlog.PositionFromStr(req.BinlogPos)
+	if err != nil {
+		return err
+	}
+
+	toInject := newOperator(req.Op, pos, events, req)
+	toInsertIndex := sort.Search(len(m.ops), func(i int) bool {
+		return pos.Compare(m.ops[i].pos) <= 0
+	})
+
+	if toInsertIndex == len(m.ops) {
+		m.ops = append(m.ops, toInject)
+		return nil
+	}
+
+	pre := m.ops[toInsertIndex]
+	if pre.pos.Compare(pos) == 0 {
+		m.ops[toInsertIndex] = toInject
+		m.logger.Warn("overwrite operator",
+			zap.Stringer("position", pos),
+			zap.Stringer("old operator", pre))
+		return nil
+	}
+
+	m.ops = append(m.ops, nil)
+	copy(m.ops[toInsertIndex+1:], m.ops[toInsertIndex:])
+	m.ops[toInsertIndex] = toInject
+	if toInsertIndex < m.nextOp {
+		m.nextOp++
+	}
+	m.logger.Info("set a new operator",
+		zap.Stringer("position", pos),
+		zap.Stringer("new operator", toInject))
+	return nil
+}
+
+// Delete will delete an operator. `posStr` should be in the format of "binlog-file:pos".
+func (m *streamModifier) Delete(posStr string) error {
+	pos, err := binlog.PositionFromStr(posStr)
+	if err != nil {
+		return err
+	}
+
+	toDeleteIndex := sort.Search(len(m.ops), func(i int) bool {
+		return pos.Compare(m.ops[i].pos) <= 0
+	})
+
+	if toDeleteIndex < m.nextOp || toDeleteIndex == len(m.ops) {
+		return terror.ErrSyncerOperatorNotExist.Generate(posStr)
+	}
+	pre := m.ops[toDeleteIndex]
+	if pre.pos.Compare(pos) != 0 {
+		return terror.ErrSyncerOperatorNotExist.Generate(posStr)
+	}
+	copy(m.ops[toDeleteIndex:], m.ops[toDeleteIndex+1:])
+	m.ops = m.ops[:len(m.ops)-1]
+	return nil
+}
+
+// ListEqualAndAfter returns a JSON string of operators equals and after the given
+// position.
+// - if argument is "", it returns all operators.
+// - Otherwise caller should make sure the argument in format of "binlog-file:pos"
+//   and it returns all operators >= this position.
+func (m *streamModifier) ListEqualAndAfter(posStr string) []*pb.HandleWorkerErrorRequest {
+	var matchedOps []*operator
+	if posStr == "" {
+		matchedOps = m.ops
+	} else {
+		pos, err := binlog.PositionFromStr(posStr)
+		if err != nil {
+			m.logger.DPanic("invalid position, should be verified in caller",
+				zap.String("position", posStr))
+			return []*pb.HandleWorkerErrorRequest{}
+		}
+
+		i := 0
+		for i < len(m.ops) && m.ops[i].pos.Compare(pos) < 0 {
+			i++
+		}
```

Comment:
why not use sort.Search too?

we can use a named method to make this common part is more clear

Reply:
https://github.com/pingcap/tiflow/pull/5890/commits/e828a097103347c27ce6d39081279901690738d8